File: app/src/main/java/com/orego/battlecrane/bc/api/context/controller/map/BMapController.java
package com.orego.battlecrane.bc.api.context.controller.map;

import com.orego.battlecrane.bc.api.context.BGameContext;
import com.orego.battlecrane.bc.api.context.controller.map.point.BPoint;
import com.orego.battlecrane.bc.api.context.eventPipeline.model.component.context.BContextComponent;
import com.orego.battlecrane.bc.api.model.contract.BUnit;
import com.orego.battlecrane.bc.api.scenario.BGameScenario;

@BContextComponent
public final class BMapController {

    public static final int MAP_SIZE = 16;

    private final BPoint[][] matrix = new BPoint[MAP_SIZE][MAP_SIZE];

    private final BGameContext context;

    public BMapController(final BGameScenario scenario, final BGameContext context) {
        for (int i = 0; i < MAP_SIZE; i++) {
            for (int j = 0; j < MAP_SIZE; j++) {
                final BPoint cell = new BPoint(i, j);
                this.matrix[i][j] = cell;
            }
        }
        scenario.initMap(this, context);
        this.context = context;
    }

    public final boolean createUnit(final BUnit unit, final BPoint pivot) {
        final int startX = pivot.getX();
        final int startY = pivot.getY();
        final int endX = startX + unit.getHorizontalSize();
        final int endY = startX + unit.getVerticalSize();
        if (this.inBounds(startX, startY)
                && this.inBounds(startX, endY)
                && this.inBounds(endX, startY)
                && this.inBounds(endX, endY)) {
            if (unit.isPlaced(this.context, pivot)) {
                //Attach pivot to entity:
                unit.setPivot(this.matrix[startX][startY]);
                //Attach entity to matrix:
                for (int i = startX; i < endX; i++) {
                    for (int j = startY; j < endY; j++) {
                        final BPoint point = BMapController.this.matrix[i][j];
                        point.setAttachedUnit(unit.getUnitId());
                    }
                }
                //Put in heap:
                final long unitId = unit.getUnitId();
                BMapController.this.unitHeap.put(unitId, unit);
                return true;
            }
        }
        return false;
    }

    public final boolean inBounds(final BPoint point) {
        return this.inBounds(point.getX(), point.getY());
    }

    public final boolean inBounds(final int x, final int y) {
        return x >= 0 && x < MAP_SIZE && y >= 0 && y < MAP_SIZE;
    }

    public final BUnit getUnitByPosition(final BPoint position) {
        return this.getUnitByPosition(position.getX(), position.getY());
    }

    public final BUnit getUnitByPosition(final int x, final int y) {
        final long unitId = this.matrix[x][y].getAttachedUnit();
        return this.unitHeap.get(unitId);
    }
}

File: app/src/main/java/com/orego/battlecrane/bc/api/context/controller/map/BMapController.kt
package com.orego.battlecrane.bc.api.context.controller.map

import com.orego.battlecrane.bc.api.context.BGameContext
import com.orego.battlecrane.bc.api.context.pipeline.model.component.context.BContextComponent
import com.orego.battlecrane.bc.api.context.storage.heap.implementation.BUnitHeap
import com.orego.battlecrane.bc.api.model.entity.main.unit.BUnit

@BContextComponent
class BMapController {

    companion object {

        const val MAP_SIZE = 16

        const val NOT_INITIALIZED_UNIT_ID: Long = -1

        fun inBounds(x: Int, y: Int) = x in 0 until MAP_SIZE && y in 0 until MAP_SIZE
    }

    private val matrix = Array(MAP_SIZE) { Array(MAP_SIZE) { NOT_INITIALIZED_UNIT_ID } }

    fun initMap(context: BGameContext){
        //Place units on mapConstraintLayout:
        val unitHeap = context.storage.getHeap(BUnitHeap::class.java)
        unitHeap.getObjectList().forEach { unit ->
            this.placeUnitOnMap(unit)
        }
        //Check initialized mapConstraintLayout:
        this.matrix.forEach { column ->
            column.forEach { id ->
                val isNotInitiablizedField = id == NOT_INITIALIZED_UNIT_ID
                if (isNotInitiablizedField) {
                    throw IllegalStateException("All mapConstraintLayout must be initialized!")
                }
            }
        }
    }

    fun placeUnitOnMap(unit: BUnit): Boolean {
        val startX = unit.x
        val startY = unit.y
        val endX = startX + unit.width
        val endY = startX + unit.height
        if (inBounds(startX, startY)
            && inBounds(startX, endY)
            && inBounds(endX, startY)
            && inBounds(endX, endY)
        ) {
            val unitId = unit.unitId
            if (unitId != NOT_INITIALIZED_UNIT_ID) {
                //Attach unit to matrix:
                for (i in startX until endX) {
                    for (j in startY until endY) {
                        this.matrix[startX][startY] = unitId
                    }
                }
                return true
            } else {
                throw IllegalStateException("Any unit must have an generated id!")
            }
        }
        return false
    }

    fun getUnitIdByPosition(x: Int, y: Int) = this.matrix[x][y]

    fun getUnitByPosition(context: BGameContext, x: Int, y: Int) =
        context.storage.getHeap(BUnitHeap::class.java)[this.getUnitIdByPosition(x, y)]
}